[feature] VQA using Large-Scale Vision and Language model

mqcmd196 opened the pull request:
The ROS package for VQA using Large scale model

### TODO
#### General
- [ ] add README and system configuration

#### OFA
- [ ] add README

knorth55 commented on jsk_vil_ros/node_scripts/ofa.py:
```diff
@@ -0,0 +1,211 @@
+    img_byte = base64.b64encode(img_str).decode("utf-8")
+    return img_byte
+
+class OfaJSKROSClient:
```
```suggestion
class OFAClientNode(object):
```

the program name should be `ofa_client_node.py`

knorth55 commented on jsk_vil_ros/apis/ofa/server.py:
```diff
@@ -0,0 +1,278 @@
+        return t.to(dtype=torch.half)
+    return t
+
+class Inference:
```
```suggestion
class Inference(object):
```

knorth55 replied:
class name should be `InferenceServer` or `HogehogeServer`.
and the program name should be `inference_server.py` or `hogehoge_server.py`

knorth55 commented on jsk_vil_msgs/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+cmake_minimum_required(VERSION 3.0.2)
+project(jsk_vil_msgs)
```
can we merge this package to `jsk_recognition_msgs?`

knorth55 commented on jsk_vil_ros/node_scripts/ofa.py:
```diff
@@ -0,0 +1,211 @@
+from sensor_msgs.msg import Image
+from std_msgs.msg import String
+
+bridge = CvBridge()
```
no global variable.

knorth55 commented on jsk_vil_ros/node_scripts/ofa.py:
```diff
@@ -0,0 +1,211 @@
+                rospy.logerr("Invalid http status code: {}".format(str(response.status_code)))
+
+    def vqa_action_cb(self, goal):
+        success = True
```
the default value should be `False` to prevent bugs.
and store `True` when it succeeded

knorth55 commented on jsk_vil_ros/package.xml:
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0"?>
+<package format="3">
+  <name>jsk_vil_ros</name>
+  <version>0.0.0</version>
```
change version number to same as other packages in `jsk_recognition`

knorth55 commented on jsk_vil_ros/package.xml:
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0"?>
```
can we merge this package to `jsk_perception`?

mqcmd196 replied:
This node is irregular in terms of setting up an API server, and at the same time, there are features in Vision and Language that are hard to classify as perceptions, such as Image Generation, so I thought it would be better to separate them. 
So, what is the best practice for `jsk_recognition` ? Adding these node files and launch files to `jsk_perception` and write docs about these nodes not working without API server ?

knorth55 replied:
i think vision and language is perception.
you can also put server and api program if you write documentation propery.

knorth55 replied:
@k-okada do you have any suggestions?

mqcmd196 replied:
Once the policy has been decided, the names of messages and packages are fixed accordingly. Some people want to use this node and I want to let them use it.

mqcmd196 replied:
I talked with @k-okada , and this PR would be merged into jsk_perception

knorth55 commented on jsk_vil_ros/apis/ofa/server.py:
```diff
@@ -0,0 +1,278 @@
+import torch
+import numpy as np
+import cv2
+from fairseq import utils, tasks
+from fairseq import checkpoint_utils
+from fairseq import options
+from fairseq.dataclass.utils import convert_namespace_to_omegaconf
+from utils.eval_utils import eval_step
+from utils.zero_shot_utils import zero_shot_step
+from tasks.mm_tasks.caption import CaptionTask
+from tasks.mm_tasks.refcoco import RefcocoTask
+from tasks.mm_tasks.vqa_gen import VqaGenTask
+from models.ofa import OFAModel
```
this writing may make people confused when you place `utils.py`, `tasks.py` and `models.py` in the same directory.
please use full path like `from fairseq.utils.eval_utils import eval_step`

Affonso-Gui commented on jsk_perception/docker/ofa/server.py:
```diff
@@ -0,0 +1,278 @@
+            data_np = np.fromstring(image_dec, dtype='uint8')
+            img = cv2.imdecode(data_np, 1)
+            # get text
+            texts = data_json['queries']
```
Shouldn't this be `questions` for the VQA? https://github.com/mqcmd196/jsk_recognition/blob/PR/large-scale-vil/jsk_recognition_msgs/action/VQATask.action#L2

```
  text: "Exception in execute callback: 'VQATaskGoal' object has no attribute 'queries'"
```

@mqcmd196

Affonso-Gui replied:
Or is the actionlib implementation deprecated and setting the dynamic reconfigure is the intended usage?

Affonso-Gui commented on jsk_perception/node_scripts/ofa_client_node.py:
```diff
@@ -0,0 +1,210 @@
+        self.host = rospy.get_param("~host", default="localhost")
+        self.port = rospy.get_param("~port", default=8888)
+        # Caption
+        self.caption_image_sub = rospy.Subscriber("~caption_image", Image,
```
Set ` self.caption_queries = None` before the subscriber? Same with `vqa_queries`.
```
Traceback (most recent call last):                                                          [231/622]
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/topics.py", line 750, in _invoke_callback 
    cb(msg)                                                                                          
  File "/home/affonso/sandbox/recognition_ws/src/jsk_recognition/jsk_perception/node_scripts/ofa_clie
nt_node.py", line 69, in caption_topic_cb                                                            
    if not self.caption_queries:                                                                     
AttributeError: 'OFAClientNode' object has no attribute 'caption_queries'  
```

k-okada commented on jsk_perception/cfg/Clip.cfg:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env python
+PACKAGE = "jsk_vil_msgs"
```
old package name?

k-okada commented on jsk_perception/docker/Makefile:
```diff
@@ -0,0 +1,43 @@
+BUILDIMAGE = docker build
+REMOVEIMAGE = docker rmi
+# configuration directories
+CONFIGDIR = $(shell echo ${HOME}/.jsk_vil_apis)
```
old package name?

k-okada commented on jsk_perception/cfg/Clip.cfg:
```diff
@@ -0,0 +1,10 @@
+
+gen.add("clip_queries", str_t, 0, "A queries of Caption Task. The sentences are separated with ;", "human;kettle;cup;glass")
+
+exit(gen.generate(PACKAGE, "vil_msgs", "ClipQuery"))
```
old package name?

k-okada commented on jsk_perception/docker/Makefile:
```diff
@@ -0,0 +1,43 @@
+ofa: download
+	$(BUILDIMAGE) $(OFAPROJECT) -t $(OFAIMAGE) -f $(OFAPROJECT)/Dockerfile
+
+all: ofa
```
did you intentionally put `all` target here? not the top of Makefile?

k-okada commented on jsk_perception/docker/Makefile:
```diff
@@ -0,0 +1,43 @@
+	@if [ ! -d $(PARAMDIR) ]; then\
+		mkdir -p $(PARAMDIR); \
+	fi
+	wget -c -i $(PARAMURLS) -P $(PARAMDIR)
```
add `touch download` in order to skip running `wget` when it already downloaded the param files. 
or set hidden file such as `$(PARAMDIR)/.download` as a make target

k-okada commented on jsk_perception/docker/ofa/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+    libopencv-dev \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+RUN git clone https://github.com/OFA-Sys/OFA \
```
Please spcify version / SHA

k-okada commented on jsk_recognition_msgs/action/ClipTask.action:
```diff
@@ -0,0 +1,7 @@
+sensor_msgs/Image image
+string[] queries
+---
+jsk_recognition_msgs/ClipResult result
```
We must be very careful on 'common noun' vs 'proper noun'
In genral, the service / task will be provided as `common noun`, because pople do not care about actual imprementation.

```
  QueryAndProbability.msg
  QuestionAndAnswerText.msg
```
This is seems good. I personally prefer `QustionAndAnswer` and `QueryAndProbability` or `QustionAndAnserText` and `QueryAndProbabilityXxxxx`.
```
  VQAResult.msg
  ClipResult.msg
```
VQA is common noun, but `Clip` is proper noun. So it should be VQAResult and VQPResult? ObjectDetetion? OpenVocavularyObjectDetection? The name of `QueryAndProbability.msg` should be update if you use other than VAP.
```
  VQATask.action
  ClipTask.action
```
Same as `Result.msg`

k-okada commented on jsk_recognition_msgs/action/ClipTask.action:
```diff
@@ -0,0 +1,7 @@
+sensor_msgs/Image image
```
should we add `CompressedImage` here?

k-okada commented on jsk_recognition_msgs/action/ClipTask.action:
```diff
@@ -0,0 +1,7 @@
+string[] queries
+---
+jsk_recognition_msgs/ClipResult result
+bool done
```
success ?

k-okada reviewed (changes requested):
I think it is almost ready to merge

- Please remove old code
- Re-consider message file name. Imagine if you have new service/network that do better job on VQA or open vocabulary object detection, how much you need to update your code.

k-okada commented on jsk_perception/docker/ofa/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+    && mkdir -p /var/mount
+RUN cd OFA \
+    && pip install -r requirements.txt \
+    && pip install "protobuf<3.20.0" "numpy>=1.20.3" flask
```
```
[ INFO] [1680156383.962963126]: [/ofa_ros_client_rviz_p51s_30801_6669198759294921931] Stereo is NOT SUPPORTED
[ INFO] [1680156383.963053325]: [/ofa_ros_client_rviz_p51s_30801_6669198759294921931] OpenGL device: Quadro P500/PCIe/SSE2
[ INFO] [1680156383.963096437]: [/ofa_ros_client_rviz_p51s_30801_6669198759294921931] OpenGl version: 4.6 (GLSL 4.6).
Traceback (most recent call last):
  File "server.py", line 7, in <module>
    from fairseq import utils, tasks
  File "/workspace/OFA/fairseq/fairseq/__init__.py", line 32, in <module>
    import fairseq.criterions  # noqa
  File "/workspace/OFA/fairseq/fairseq/criterions/__init__.py", line 36, in <module>
    importlib.import_module("fairseq.criterions." + file_name)
  File "/opt/conda/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/workspace/OFA/fairseq/fairseq/criterions/ctc.py", line 18, in <module>
    from fairseq.data.data_utils import post_process
  File "/workspace/OFA/fairseq/fairseq/data/__init__.py", line 24, in <module>
    from .indexed_dataset import (
  File "/workspace/OFA/fairseq/fairseq/data/indexed_dataset.py", line 121, in <module>
    6: np.float,
  File "/opt/conda/lib/python3.8/site-packages/numpy/__init__.py", line 305, in __getattr__
    raise AttributeError(__former_attrs__[attr])
AttributeError: module 'numpy' has no attribute 'float'.
`np.float` was a deprecated alias for the builtin `float`. To avoid this error in existing code, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
[ofa_api-2] process has finished cleanly
log file: /home/k-okada/.ros/log/fda89ed2-cec0-11ed-96cb-2b147920af86/ofa_api-2*.log
^C[ofa_ros_client_rqt_p51s_30801_7031696000715711632-5] killing on exit
[ofa_ros_client_rviz_p51s_30801_6669198759294921931-4] killing on exit
```
need to set `numpy>=1.20.3,<1.24.0` ? https://stackoverflow.com/questions/74844262/how-can-i-solve-error-module-numpy-has-no-attribute-float-in-python

k-okada replied:
```
  File "/opt/conda/lib/python3.8/site-packages/lightning_fabric/__init__.py", line 29, in <module>
    from lightning_fabric.fabric import Fabric  # noqa: E402
  File "/opt/conda/lib/python3.8/site-packages/lightning_fabric/fabric.py", line 32, in <module>
    from lightning_fabric.plugins import Precision  # avoid circular imports: # isort: split
  File "/opt/conda/lib/python3.8/site-packages/lightning_fabric/plugins/__init__.py", line 18, in <module>
    from lightning_fabric.plugins.precision.amp import MixedPrecision
  File "/opt/conda/lib/python3.8/site-packages/lightning_fabric/plugins/precision/__init__.py", line 14, in <module>
    from lightning_fabric.plugins.precision.amp import MixedPrecision
  File "/opt/conda/lib/python3.8/site-packages/lightning_fabric/plugins/precision/amp.py", line 29, in <module>
    class MixedPrecision(Precision):
  File "/opt/conda/lib/python3.8/site-packages/lightning_fabric/plugins/precision/amp.py", line 96, in MixedPrecision
    def _autocast_context_manager(self) -> torch.autocast:
AttributeError: module 'torch' has no attribute 'autocast'
[ofa_api-2] process has finished cleanly

```
```
  File "/opt/conda/lib/python3.8/site-packages/datasets/arrow_dataset.py", line 64, in <module>
    from . import config
  File "/opt/conda/lib/python3.8/site-packages/datasets/config.py", line 134, in <module>
    importlib.import_module("soundfile").__libsndfile_version__
  File "/opt/conda/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/opt/conda/lib/python3.8/site-packages/soundfile.py", line 192, in <module>
    _snd = _ffi.dlopen(_explicit_libname)
OSError: cannot load library 'libsndfile.so': libsndfile.so: cannot open shared object file: No such file or directory
[ofa_api-2] process has finished cleanly
```

```
    from .speech_tasks import *
  File "/workspace/OFA/tasks/speech_tasks/__init__.py", line 1, in <module>
    from .unify_task import UnifySpeechTextTask
  File "/workspace/OFA/tasks/speech_tasks/unify_task.py", line 16, in <module>
    from data.s2t_data.unify_dataset import UnifyDataset
  File "/workspace/OFA/data/s2t_data/unify_dataset.py", line 33, in <module>
    import torchaudio
ModuleNotFoundError: No module named 'torchaudio'
[ofa_api-2] process has finished cleanly
```
```
  File "server.py", line 11, in <module>
    from utils.eval_utils import eval_step
  File "/workspace/OFA/utils/eval_utils.py", line 17, in <module>
    from tasks.nlg_tasks.gigaword import fix_tokenization
  File "/workspace/OFA/tasks/__init__.py", line 6, in <module>
    from .speech_tasks import *
  File "/workspace/OFA/tasks/speech_tasks/__init__.py", line 1, in <module>
    from .unify_task import UnifySpeechTextTask
  File "/workspace/OFA/tasks/speech_tasks/unify_task.py", line 16, in <module>
    from data.s2t_data.unify_dataset import UnifyDataset
  File "/workspace/OFA/data/s2t_data/unify_dataset.py", line 38, in <module>
    from g2p_en import G2p
ModuleNotFoundError: No module named 'g2p_en'
[ofa_api-2] process has finished cleanly
```

mqcmd196 commented:
requirements.txt
```
absl-py==1.2.0
aiohttp==3.8.3
aiosignal==1.2.0
antlr4-python3-runtime==4.8
async-timeout==4.0.2
attrs==22.1.0
backcall @ file:///home/ktietz/src/ci/backcall_1611930011877/work
beautifulsoup4 @ file:///home/linux1/recipes/ci/beautifulsoup4_1610988766420/work
bitarray==2.6.0
brotlipy==0.7.0
cachetools==5.2.0
certifi==2020.12.5
cffi @ file:///tmp/build/80754af9/cffi_1605538068321/work
chardet @ file:///tmp/build/80754af9/chardet_1605303185383/work
charset-normalizer==2.1.1
click==8.1.3
colorama==0.4.5
conda==4.9.2
conda-build==3.21.4
conda-package-handling @ file:///tmp/build/80754af9/conda-package-handling_1603018141399/work
contourpy==1.0.5
cryptography @ file:///tmp/build/80754af9/cryptography_1605544487601/work
cycler==0.11.0
Cython==0.29.32
datasets==2.5.1
decorator @ file:///home/ktietz/src/ci/decorator_1611930055503/work
dill==0.3.5.1
dnspython==2.1.0
einops==0.4.1
-e git+https://github.com/OFA-Sys/OFA@1809b55687458dbb4b1686a89b569f0efb2536e3#egg=fairseq&subdirectory=fairseq
filelock @ file:///home/linux1/recipes/ci/filelock_1610993975404/work
Flask==2.2.2
fonttools==4.37.4
frozenlist==1.3.1
fsspec==2022.8.2
ftfy==6.0.3
glob2 @ file:///home/linux1/recipes/ci/glob2_1610991677669/work
google-auth==2.12.0
google-auth-oauthlib==0.4.6
grpcio==1.49.1
huggingface-hub==0.10.0
hydra-core==1.0.7
idna @ file:///tmp/build/80754af9/idna_1593446292537/work
importlib-metadata==5.0.0
importlib-resources==5.9.0
ipython @ file:///tmp/build/80754af9/ipython_1614615245721/work
ipython-genutils @ file:///tmp/build/80754af9/ipython_genutils_1606773439826/work
itsdangerous==2.1.2
jedi==0.17.0
Jinja2==3.1.2
joblib==1.2.0
kiwisolver==1.4.4
libarchive-c @ file:///home/linux1/recipes/ci/python-libarchive-c_1610974153025/work
lxml==4.9.1
Markdown==3.4.1
MarkupSafe==2.1.1
matplotlib==3.6.0
mkl-fft==1.3.0
mkl-random==1.1.1
mkl-service==2.3.0
multidict==6.0.2
multiprocess==0.70.13
nltk==3.7
numpy==1.23.3
oauthlib==3.2.1
olefile==0.46
omegaconf==2.0.6
opencv-python==4.6.0.66
packaging==21.3
pandas==1.5.0
parso @ file:///tmp/build/80754af9/parso_1607623074025/work
pexpect @ file:///tmp/build/80754af9/pexpect_1605563209008/work
pickleshare @ file:///tmp/build/80754af9/pickleshare_1606932040724/work
Pillow @ file:///tmp/build/80754af9/pillow_1615224089586/work
pkginfo==1.7.0
portalocker==2.5.1
prompt-toolkit @ file:///tmp/build/80754af9/prompt-toolkit_1602688806899/work
protobuf==3.19.6
psutil @ file:///tmp/build/80754af9/psutil_1612298023621/work
ptyprocess @ file:///tmp/build/80754af9/ptyprocess_1609355006118/work/dist/ptyprocess-0.7.0-py2.py3-none-any.whl
pyarrow==9.0.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycocoevalcap==1.2
pycocotools==2.0.4
pycosat==0.6.3
pycparser @ file:///tmp/build/80754af9/pycparser_1594388511720/work
pyDeprecate==0.3.2
Pygments @ file:///tmp/build/80754af9/pygments_1613849567472/work
pyOpenSSL @ file:///tmp/build/80754af9/pyopenssl_1605545627475/work
pyparsing==3.0.9
PySocks @ file:///tmp/build/80754af9/pysocks_1605305779399/work
python-dateutil==2.8.2
python-etcd==0.4.5
pytorch-lightning==1.7.7
pytz @ file:///tmp/build/80754af9/pytz_1612215392582/work
PyYAML==5.4.1
regex==2022.9.13
requests @ file:///tmp/build/80754af9/requests_1592841827918/work
requests-oauthlib==1.3.1
responses==0.18.0
rouge-score==0.1.2
rsa==4.9
ruamel-yaml==0.15.87
sacrebleu==2.2.1
six @ file:///tmp/build/80754af9/six_1605205327372/work
soupsieve @ file:///tmp/build/80754af9/soupsieve_1614023247537/work
tabulate==0.8.10
tensorboard==2.10.1
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
tensorboardX==2.4.1
timm==0.6.7
torch==1.8.1
torchelastic==0.2.2
torchmetrics==0.9.3
torchtext==0.9.1
torchvision==0.9.1
tqdm @ file:///tmp/build/80754af9/tqdm_1605303662894/work
traitlets @ file:///home/ktietz/src/ci/traitlets_1611929699868/work
typing-extensions==4.3.0
urllib3 @ file:///tmp/build/80754af9/urllib3_1603305693037/work
wcwidth @ file:///tmp/build/80754af9/wcwidth_1593447189090/work
Werkzeug==2.2.2
xxhash==3.0.0
yarl==1.8.1
zipp==3.8.1
```

mqcmd196 commented:
@k-okada 
As yesterday's discussion, I reworked VQA node.

https://github.com/jsk-ros-pkg/jsk_recognition/pull/2730/commits/693fd1c6a7ade9789172b75af4bee308f57fa535, https://github.com/jsk-ros-pkg/jsk_recognition/pull/2730/commits/543276960214e0a2f0247ac9207753534b054156
- removed CaptionAction and merged into VQA. (use OFA's caption backend)

https://github.com/jsk-ros-pkg/jsk_recognition/pull/2730/commits/df021248832a5f0316866992f79b28949a9ae10a
- add sample launch

tkmtnt7000 commented on jsk_perception/launch/classification.launch:
```diff
@@ -0,0 +1,22 @@
+  <node name="classification_api" pkg="jsk_perception" type="run_jsk_vil_api" output="log"
+        args="clip -p $(arg port)" if="$(arg run_api)" />
+
+  <node name="classification" pkg="jsk_perception" type="classification_client_node.py" output="screen">
```
??
```suggestion
  <node name="classification" pkg="jsk_perception" type="classification_node.py" output="screen">
```

mqcmd196 replied:
Thank you, fixed

tkmtnt7000 commented on jsk_perception/node_scripts/classification_node.py:
Please `chmod u+x` to execute from `rosrun`

mqcmd196 commented:
@k-okada 
I merged both ofa, clip client into single file
https://github.com/jsk-ros-pkg/jsk_recognition/pull/2730/commits/138ec46e155bbd0974ce0676fab79c10ee192570, https://github.com/jsk-ros-pkg/jsk_recognition/pull/2730/commits/aec353601c9b17841983251fa7ba045e3f98a180, https://github.com/jsk-ros-pkg/jsk_recognition/pull/2730/commits/356802edfac7c29ae82878edcc70bf9481b6c818, https://github.com/jsk-ros-pkg/jsk_recognition/pull/2730/commits/3e059b85b677188c5e2cf3c991d37792ca12d198

https://github.com/jsk-ros-pkg/jsk_recognition/pull/2730/files#diff-6d0168ccff6ccc2f731c899cb021a76d1c315ac0c28a874a259c80ceabf994b6

k-okada commented:
@mqcmd196 ping...

mqcmd196 commented:
I'm sorry... updated

tkmtnt7000 commented on doc/jsk_perception/nodes/vqa_node.md:
```diff
@@ -0,0 +1,91 @@
+The ROS node for Visual Question Answering(VQA).
+
+## System Configuration
+![][images/large_scale_vil_system.pdf]
```
Same as above.

If we want to show a system configutation figure directly on Readme.md, we should replace pdf file to some other image file formats (jpeg or png).

tkmtnt7000 reviewed (changes requested):
Test passed and I think we can almost merge them if something little has been revised.

mqcmd196 commented:
Thank you, updated
